lib/Transforms/LayoutOptimization/LayoutOptimization.cpp

```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
+                                                      kernel::SymbolicValue>>) {
+            return processInternal(arg.operand) + 1;
+          } else if constexpr (std::is_same_v<
+                                   T, kernel::ExtractNode<kernel::SymbolicValue>>) {
+            return processInternal(arg.operand);
+          }
+          return 0;
+        },
+        node->node_variant);
+  }
+};
+
+// Helper: Extract operation shape for kernel cost calculation
+static std::optional<ArrayRef<int64_t>> getOperationShape(Operation* op) {
+  return llvm::TypeSwitch<Operation*, std::optional<ArrayRef<int64_t>>>(op)
+      .Case<linalg::MatvecOp>(
+          [](auto matvecOp) -> std::optional<ArrayRef<int64_t>> {
+            // Matrix is operand 0: rows x cols
+            auto matrixType =
+                dyn_cast<RankedTensorType>(matvecOp.getInputs()[0].getType());
+            if (!matrixType) return std::nullopt;
+            return matrixType.getShape();
+          })
+      .Case<linalg::VecmatOp>(
+          [](auto vecmatOp) -> std::optional<ArrayRef<int64_t>> {
+            // Matrix is operand 1: rows x cols
+            auto matrixType =
+                dyn_cast<RankedTensorType>(vecmatOp.getInputs()[1].getType());
+            if (!matrixType) return std::nullopt;
+            return matrixType.getShape();
+          })
+      .Case<linalg::MatmulOp>(
+          [](auto matmulOp) -> std::optional<ArrayRef<int64_t>> {
+            // LHS matrix is operand 0
+            auto lhsType =
+                dyn_cast<RankedTensorType>(matmulOp.getInputs()[0].getType());
+            if (!lhsType) return std::nullopt;
+            return lhsType.getShape();
+          })
+      .Default([](auto) { return std::nullopt; });
+}
+
+// Helper: Build symbolic DAG for a kernel and count rotations
+static Cost computeKernelCostFromDAG(KernelName kernel,
+                                     ArrayRef<int64_t> shape) {
+  using NodeTy = kernel::ArithmeticDagNode<kernel::SymbolicValue>;
+  std::shared_ptr<NodeTy> kernelDag;
+
+  switch (kernel) {
+    case KernelName::Trivial:
+      return 0;
+
+    case KernelName::MatvecDiagonal: {
+      if (shape.size() < 2) return 0;
+      // Create symbolic matrix (rows x cols) and vector (cols)
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      kernelDag =
+          kernel::implementMatvec(kernel, symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::MatvecNaive: {
+      if (shape.size() < 2) return 0;
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      // MatvecNaive not yet implemented in KernelImplementation.h
+      // For now, use diagonal as approximation
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
```

Comment:
Better is to return a `FailureOr<Cost>` and fail here, let the caller decide how to handle it (perhaps, return a very large sentinel cost or abort).